Add PowerSGD optimizer

mmilenkoski opened the pull request:
The code is adapted from [PowerSGD](https://github.com/epfml/powersgd)

martinjaggi commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+            return buffers
+
+
+def all_reduce(*args, **kwargs):
```
should document somewhere which kind of communication backend and custom `all_reduce` we're supporting for this one (and explain a bit the difference to the standard reference implementations in that respect).

now it's not clear to the user if this all_reduce here is supposed to be used by all optimizers or just by powerSGD. or would it be simpler to make a new file `optim_powerSGD.py` or similar, to avoid confusion for SGD people who don' t need this one?

mmilenkoski replied:
Maybe it would be enough if I just move the functions inside the PowerSGD class? That way it is understood that they are used only by that class.

ehoelzl replied:
I don't really see the use case for this function. How is it different than using `dist.all_reduce` ? We are always initializing the distributed environment. Can you post an example where this would be useful ?

tvogels replied:
I think that this class is too heavy. It currently serves more than one purposes: packing/unpacking + communication. I recently started using this snippet for packing only, which I find cleaner:

```
def pack(tensors):
    """Packs a list of tensors into one buffer for sending to other workers"""
    buffer = torch.cat([t.view(-1) for t in tensors])  # copies
    shapes = [tensor.shape for tensor in tensors]
    return buffer, shapes


def unpack(buffer, shapes):
    """Provides pointers to tensors of original `shapes` in a flat-packed buffer."""
    idx = 0
    entries = []
    for tensor_shape in shapes:
        end = idx + tensor_shape.numel()
        entries.append(buffer[idx:end].view(*tensor_shape))
        idx = end

    return entries
```

martinjaggi commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+
+
+@torch.jit.script
+def orthogonalize(matrix):
```
comment that needed only for powerSGD

ehoelzl replied:
Maybe move this to a utility module ? Maybe create a file `mlbench_core/optim/pytorch/utils.py`, as we will probably need it later.

martinjaggi commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def all_gather(out_list, in_tensor, **kwargs):
```
same comment as above. is this one for all optimizers or not

ehoelzl replied:
Same comment as above.

Also, if we need to keep them, they can be moved to `mlbench_core/utils/pytorch/distributed.py`

martinjaggi commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+        nesterov=False,
+        reuse_query=False,
+        rank=1,
+    ):
```
looks nice! in other places where rank is used for a different meaning, comment to clarify. and here clarify that its matrix approximation rank

mmilenkoski replied:
In the other classes, the meaning of `rank` is explained inside a docstring. I will do that also for this class.

Panaetius replied:
Please always add google style docstrings to all non-private methods at the least and ideally to all methods. http://google.github.io/styleguide/pyguide.html#383-functions-and-methods

martinjaggi replied:
documentation comment: could someone make a separate issue to more clearly document node id/rank parameters as "unique id/rank of this node" in all other mlbench places. this would avoid confusion with low-rank in the linear algebra sense here

martinjaggi commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+        rank1_tensor_list.unpack([out for (_, out, _) in rank1_tensors])
+
+
+class TensorBuffer:
```
should this one be used by the all optimizers?

mmilenkoski replied:
No, only by PowerSGD. I will put a docstring to explain this.

martinjaggi replied:
it's related to the per-layer or per-model aggregation also, which we are supporting for all optimizers #61 . maybe the code could be shared?

ehoelzl replied:
Ideally, this class should be moved to `mlbench_core/utils/pytorch/distributed.py` since it doesn't really have to do with optimizers. Also, a function `pack_tensors` already exists, and I'm not sure we need a whole class for it. Please check how it's used.

ehoelzl commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+        dampening=0,
+        nesterov=False,
+        reuse_query=False,
+        rank=1,
```
Shouldn't the default rank be 0 ?

martinjaggi replied:
0 would transmit nothing. 1 is the correct default for low-rank compression

ehoelzl commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
         return loss
 
 
+class PowerSGD(Optimizer):
```
Great work. One small comment: Is it possible to separate the aggregation scheme from the Optimizer itself ? Like instead of defining an `all_reduce`, maybe create a new aggregator ? Would be useful in case we want to re-use it for different optimizers.

tvogels replied:
I think that those `all_reduce` and `all_gather` functions in the global scope can be removed, but I don't see why you would want to use other aggregators than `torch.distributed.all_reduce`. Do you already have other instances of that in the codebase?

mmilenkoski replied:
@ehoelzl Did you mean moving the `reduce` function from PowerSGD to a new aggregator?

tvogels commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,302 @@ def step(self, closure=None):
+    ):
+        if not model:
+            raise ValueError('"model" not set for optimizer')
+        if lr is not required and lr < 0.0:
```
I think it would be nicer to separate the optimizer and the aggregator (PowerSGD w/ EF), so you can plug it in instead of all-reduce into any optimizer (Adam, Adagrad, ...)

tvogels commented on mlbench_core/optim/pytorch/optim.py:
```diff
@@ -499,12 +499,302 @@ def step(self, closure=None):
+            rank = min(n, m, self.rank)
+            ps.append(self.p_memory[p_idx : p_idx + n * rank].view(n, rank))
+            qs.append(self.q_memory[q_idx : q_idx + m * rank].view(m, rank))
+            p_idx += n * rank
```
Maybe it would be nice to abstract away this bookkeeping into a `pack` and `unpack` function (that operates without copying), if there is not already something like that in the codebase.

mmilenkoski commented:
I believe I addressed all of the comments with the newest version. I had to override the function `_agg_gradients_by_model` in the new `PowerAggregation` class due to the way that aggregation is done in PowerSGD. I am not sure whether it makes sense (and whether we want) to allow aggregation by layer and aggregation of weights in the new aggregator.

tvogels commented on mlbench_core/utils/pytorch/utils.py:
```diff
@@ -0,0 +1,16 @@
+    for i in range(m):
+        # Normalize the i'th column
+        col = matrix[:, i : i + 1]
+        col /= torch.sqrt(torch.sum(col ** 2))
```
This could lead to division by zero. I'd suggest to add a `+ eps` here, which is configurable as a second argument and has a default of `eps=1e-16`. I'm not so familiar with the project yet, but if there are unit tests, would be good to add some here (columns have unit length & columns have zero dot product)

mmilenkoski replied:
Sure, I will add this. FYI, I copied the function from [here](https://github.com/epfml/powersgd/blob/3f4eb8d097e9a25ba4fd4b2ff1b2bf83c5aa665a/gradient_reducers.py#L945), so maybe it should be changed there too.

tvogels commented:
Looks good to me! Have you already tested the algorithm and verified you get similar test accuracy as the original code on Cifar 10 / ResNet 18?
![image](https://user-images.githubusercontent.com/840825/80341995-b8b17980-8863-11ea-9b37-108e516ad305.png)
with these [hyperparameters](https://github.com/epfml/powersgd/blob/master/hyperparameters.md#shared-parameters):
16 workers, lr schedule: 0.1 -> 1.6 in 5 epochs (step-wise), decay to 0.16 from epoch 150, 0.016 from epoch 250, weight decay 1e-4, 0 for batch norm layers, batch size 128 per worker, nesterov momentum of 0.9, with error feedback (memory), `optimizer_reducer_reuse_query` (warm start) enabled

mmilenkoski commented:
> Looks good to me! Have you already tested the algorithm and verified you get similar test accuracy as the original code on Cifar 10 / ResNet 18?
> ![image](https://user-images.githubusercontent.com/840825/80341995-b8b17980-8863-11ea-9b37-108e516ad305.png)
> with these [hyperparameters](https://github.com/epfml/powersgd/blob/master/hyperparameters.md#shared-parameters):
> 16 workers, lr schedule: 0.1 -> 1.6 in 5 epochs (step-wise), decay to 0.16 from epoch 150, 0.016 from epoch 250, weight decay 1e-4, 0 for batch norm layers, batch size 128 per worker, nesterov momentum of 0.9, with error feedback (memory), `optimizer_reducer_reuse_query` (warm start) enabled

@martinjaggi should I test the optimizer with these hyperparameters or do we want to do some standardized comparison between the optimizers?

tvogels commented:
> @martinjaggi should I test the optimizer with these hyperparameters or do we want to do some standardized comparison between the optimizers?

In the end the goal is to run the standardised comparisons. Running the settings from the paper is just interesting once to validate the correctness of the code.
